### PR TITLE
Next round of updates for fusion heuristics

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Issue #9259 
Issue #9139 
Issue #9303 
Issue #8978 
Issue #9772

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
@pzread This is the tip of changes I have been working on. 
Note that there are a couple of LLVM changes needed here. I have pushed those changes to https://github.com/iree-org/iree-llvm-fork/tree/mahesh/multi_result_elementwise and those commits are pulled in here.
One of the changes to LLVM (https://github.com/iree-org/iree-llvm-fork/commit/d7e3605316e735d8efcfcebe973a315f549bcf3c) is ready to land. Ill land that and associated changes in this branch to IREE main and rebase this branch on top of those changes.

Current status: 

The first patch in this disables the reduction -> bcast -> elementwise fusion added to IREE on main. I found it hard to adapt and need to revisit that. 

At head the dispatch statistics is 

| Model | Fills | Copies | Dispatches | Executables | Generic Before | Generic After | DispatchWorkgroups |
| --- | --- | --- | --- | --- | --- | --- | --- |
| bert_encoderhead | 12 | 4 | 537 | 41 | 2059 | 485 | 561 |
| collatzhead | 2 | 1 | 0 | 41 | 12 | 0 | 0 |
| deeplabhead | 19 | 1 | 79 | 41 | 106 | 59 | 79 |
| edge_detectionhead | 1 | 0 | 3 | 41 | 4 | 1 | 4 |
| efficientnethead | 17 | 1 | 121 | 41 | 213 | 85 | 121 |
| fragmenthead | 3 | 1 | 1 | 41 | 11 | 1 | 1 |
| fullyconnectedhead | 0 | 0 | 3 | 41 | 12 | 3 | 3 |
| minilmhead | 0 | 2 | 221 | 41 | 1329 | 248 | 221 |
| mnisthead | 4 | 0 | 6 | 41 | 10 | 6 | 6 |
| mobilebert-fp32head | 49 | 196 | 752 | 41 | 1765 | 652 | 752 |
| mobilebert-int8head | 50 | 196 | 1472 | 41 | 4773 | 1372 | 1472 |
| mobilenetv2head | 18 | 1 | 74 | 41 | 103 | 56 | 74 |
| mobilenetv3head | 12 | 1 | 96 | 41 | 226 | 85 | 96 |
| mobilessdhead | 36 | 16 | 121 | 41 | 157 | 88 | 126 |
| persondetect-int8head | 14 | 1 | 75 | 41 | 95 | 45 | 75 |
| posenethead | 15 | 1 | 48 | 41 | 59 | 32 | 48 |
| resnet50head | 346 | 0 | 78 | 41 | 741 | 55 | 78 |
| unidirectional_lstmhead | 7 | 9 | 11 | 41 | 43 | 8 | 11 |

Disabling that change for now leads to number of dispatches being

| Model | Fills | Copies | Dispatches | Executables | Generic Before | Generic After | DispatchWorkgroups |
| --- | --- | --- | --- | --- | --- | --- | --- |
| bert_encoderbase | 12 | 4 | 585 | 41 | 2059 | 485 | 609 |
| collatzbase | 2 | 1 | 0 | 41 | 12 | 0 | 0 |
| deeplabbase | 19 | 1 | 79 | 41 | 106 | 59 | 79 |
| edge_detectionbase | 1 | 0 | 3 | 41 | 4 | 1 | 4 |
| efficientnetbase | 17 | 1 | 121 | 41 | 213 | 85 | 121 |
| fragmentbase | 3 | 1 | 1 | 41 | 11 | 1 | 1 |
| fullyconnectedbase | 0 | 0 | 3 | 41 | 12 | 3 | 3 |
| minilmbase | 0 | 2 | 271 | 41 | 1329 | 248 | 271 |
| mnistbase | 4 | 0 | 6 | 41 | 10 | 6 | 6 |
| mobilebert-fp32base | 49 | 196 | 752 | 41 | 1765 | 652 | 752 |
| mobilebert-int8base | 50 | 196 | 1472 | 41 | 4773 | 1372 | 1472 |
| mobilenetv2base | 18 | 1 | 74 | 41 | 103 | 56 | 74 |
| mobilenetv3base | 12 | 1 | 96 | 41 | 226 | 85 | 96 |
| mobilessdbase | 36 | 16 | 121 | 41 | 157 | 88 | 126 |
| persondetect-int8base | 14 | 1 | 75 | 41 | 95 | 45 | 75 |
| posenetbase | 15 | 1 | 48 | 41 | 59 | 32 | 48 |
| resnet50base | 346 | 0 | 78 | 41 | 741 | 55 | 78 |
| unidirectional_lstmbase | 7 | 9 | 11 | 41 | 43 | 8 | 11 |

So number of dispatches in  miniLM increases. For now use this as a baseline. Eventually we need to get back on par with HEAD.

With the tip of this branch I get 

| Model | Fills | Copies | Dispatches | Executables | Generic Before | Generic After | DispatchWorkgroups |
| --- | --- | --- | --- | --- | --- | --- | --- |
| bert_encodermodified_1 | 12 | 4 | 561 | 41 | 2059 | 437 | 585 |
| collatzmodified_1 | 2 | 1 | 0 | 41 | 12 | 0 | 0 |
| deeplabmodified_1 | 19 | 1 | 79 | 41 | 106 | 59 | 79 |
| edge_detectionmodified_1 | 1 | 0 | 3 | 41 | 4 | 1 | 4 |
| efficientnetmodified_1 | 17 | 1 | 120 | 41 | 213 | 84 | 120 |
| fragmentmodified_1 | 3 | 1 | 1 | 41 | 11 | 1 | 1 |
| fullyconnectedmodified_1 | 0 | 0 | 3 | 41 | 12 | 3 | 3 |
| minilmmodified_1 | 0 | 2 | 258 | 41 | 1329 | 174 | 258 |
| mnistmodified_1 | 4 | 0 | 6 | 41 | 10 | 6 | 6 |
| mobilebert-fp32modified_1 | 49 | 196 | 751 | 41 | 1765 | 651 | 751 |
| mobilebert-int8modified_1 | 50 | 196 | 1447 | 41 | 4773 | 1347 | 1447 |
| mobilenetv2modified_1 | 18 | 1 | 74 | 41 | 103 | 56 | 74 |
| mobilenetv3modified_1 | 12 | 1 | 96 | 41 | 226 | 85 | 96 |
| mobilessdmodified_1 | 36 | 16 | 121 | 41 | 157 | 88 | 126 |
| persondetect-int8modified_1 | 14 | 1 | 74 | 41 | 95 | 44 | 74 |
| posenetmodified_1 | 15 | 1 | 48 | 41 | 59 | 32 | 48 |
| resnet50modified_1 | 346 | 0 | 78 | 41 | 741 | 55 | 78 |
| unidirectional_lstmmodified_1 | 7 | 9 | 9 | 41 | 43 | 6 | 9 |

Most benefits seem to be on bert based models only....

**Comment by MaheshRavishankar** [comment]
The next step here is to incorporate this https://github.com/MaheshRavishankar/iree/commit/ddeacc182482294a3a25ee5d88ea5d32a792dc8a patch onto this branch

**Comment by MaheshRavishankar** [comment]
https://reviews.llvm.org/D132716 upstreams all the required MLIR changes.

**Comment by powderluv** [comment]
how does the increase in miniLM dispatches from 221 to 271 affect performance on CPU and gpu ?  We need to ideally get to 179 dispatches or less.

**Comment by MaheshRavishankar** [comment]
Eventually it will decrease. I found it hard to generalize the existing heuristic that was added recently to get it down. So I disabled that. That's why the number of dispatches increased. After everything works it should cover those cases as well. 

For the example, the tip of this branch fuses softmax into a dispatch. Also fuses batchnorm training into a single dispatch. These are all strict improvements. What is listed above is with just after 1/3rd of the work. 

There are issues in the backend that I am fixing. Will have updated numbers after that. 
W.r.t to performance, the GPU backend might need work to get performance back up even if the number of dispatches reduce. Right now I am focusing on CPU backends. 

(Also this marked as a draft, I obviously won't be landing something that regresses performance)

**Comment by powderluv** [comment]
no worries. I know this is a hard problem to manage all of the cases. I think that is ok to regress in the short term to get better in the long term but ideally if there is some knob for the best perf case would be good to have that for benchmarks etc.

**Comment by MaheshRavishankar** [comment]
Yeah, if I hit more issues, I will land it behind a flag to not regress performance.

**Comment by MaheshRavishankar** [comment]
The heuristics in this PR now create the following number of dispatches 

| Model | Fills | Copies | Dispatches | Executables | Generic Before | Generic After | DispatchWorkgroups |
| --- | --- | --- | --- | --- | --- | --- | --- |
| bert_encodermodified | 12 | 4 | 513 | 41 | 2059 | 437 | 537 |
| collatzmodified | 2 | 1 | 0 | 41 | 12 | 0 | 0 |
| deeplabmodified | 19 | 1 | 79 | 41 | 106 | 59 | 79 |
| edge_detectionmodified | 1 | 0 | 3 | 41 | 4 | 1 | 4 |
| efficientnetmodified | 17 | 1 | 120 | 41 | 213 | 84 | 120 |
| fragmentmodified | 3 | 1 | 1 | 41 | 11 | 1 | 1 |
| fullyconnectedmodified | 0 | 0 | 3 | 41 | 12 | 3 | 3 |
| minilmmodified | 0 | 2 | 184 | 41 | 1329 | 174 | 184 |
| mnistmodified | 4 | 0 | 4 | 41 | 10 | 6 | 4 |
| mobilebert-fp32modified | 49 | 196 | 727 | 41 | 1765 | 651 | 727 |
| mobilebert-int8modified | 50 | 196 | 1403 | 41 | 4773 | 1347 | 1403 |
| mobilenetv2modified | 18 | 1 | 73 | 41 | 103 | 56 | 73 |
| mobilenetv3modified | 12 | 1 | 96 | 41 | 226 | 85 | 96 |
| mobilessdmodified | 36 | 16 | 121 | 41 | 157 | 88 | 126 |
| persondetect-int8modified | 14 | 1 | 74 | 41 | 95 | 44 | 74 |
| posenetmodified | 15 | 1 | 48 | 41 | 59 | 32 | 48 |
| resnet50modified | 346 | 0 | 76 | 41 | 741 | 55 | 76 |
| unidirectional_lstmmodified | 7 | 9 | 9 | 41 | 43 | 6 | 9 |

@powderluv the MiniLM numbers might be of interest to you. The number of dispatches is 184.

The PR is currently blocked by backend issues. On CPU side the major blocker is https://github.com/llvm/llvm-project/issues/57205 . https://reviews.llvm.org/D133474 should address this, but there is a long road to get this landed and used in IREE.

**Comment by dcaballe** [comment]
Please, keep me posted on the landing. This will impact all the ongoing perf analysis.

**Comment by powderluv** [comment]
This is great work. Can't wait to re-run all our perf benchmarks.

**Comment by MaheshRavishankar** [comment]
Subsumed by #10472
